# Как понимать код на Python

## Язык с интуитивно понятным синтаксисом

Синтаксис языка Python очень интуитивен и понятен сам по себе. Когда говорят "синтаксис", имеют ввиду используемый в языке набор ключевых слов, правил их использования и правил оформления кода в целом. Именно с изучения синтаксиса языка как правило и начинают знакомиться с новым языком.

Как первый язык программирования Python тоже довольно хорош за счет этой интуитивности, хотя все зависит от ваших целей. Если вы собираетесь в последствии заняться серьезной разработкой программного обеспечения на более сложных для освоения языках - например, на C++ или Java, то легкость и интуитивность Python может наоборот оказать "расслабляющее" действие и после него учить сложные языки будет тяжелее. Но совершенно точно Python дает очень быстрый путь от начала обучения, до умения применить программирование с пользой.

Посмотрим, например, как выглядит программа "Hello, World!" на Python и C++. По традиции с этой программы начинают знакомство с языками программирования, при этом все что нужно сделать - вывести на экран строчку "Hellow, World!" ("Привет, мир!")

*Python:*

In [1]:
print("Hello, World!")

Hello, World!


*C++:*

In [ ]:
#include <iostream>

void main() {
    std::cout << "Hello, World";
}

Или, еще непонятней, но вполне корректно, так:

In [ ]:
#include <iostream>

int main() {
    std::cout << "Hello, World";
    return 0;
}

И это лучший момент для драматичной паузы :)

Если в Python мы дословно пишем: "напечатать" (print) строку "Hello, World!", то в C++ сначала подключаем какую-то библиотеку iostream, потом пишем какой-то main (это функция, которая выполняется при запуске программы в C++, исполняемый код пишется в ней, потом пишем какое-то шаманское "std::" (обращение к так называемой "стандартной библиотеке"), потом "cout" (стандартный поток вывода), потом "<<" (на самом деле в С++ это оператор побитового сдвига, но также еще этот же оператор используется для перенаправления вывода в стандартный поток) и потом уже наконец строчка "Hello, World!". Ужас!

В итоге, когда вы видите код на Python, нужно только сообразить, что инструкция "вывести на печать" дальше в скобочках принимает то, что нужно выводить на печать, а также, что текстовые строки в Python пишутся в кавычках. А в C++ надо либо принять все эти волшебные заклинания как данность и понять их когда-нибудь по-позже (плохой вариант), либо разобраться, что такое стандартная библиотека, стандартный поток вывода, перенаправление в поток, какой может быть функция main и что она может возвращать.

Т.к. цель - рассказать по Python, я не буду больше углубляться в пример на C++ и не раскрою смысл всех этих умных слов, хотя, если вам стало интересно - могу порекомендовать попробовать курсы Яндекса по C++ на Coursera.org

Но этот пример позволяет нам сделать основное предположение насчет Python: код на нем настолько понятен, что часто с ним можно неплохо разбираться и не зная Python вовсе! И даже если вы не программист. Об этом мы и начнем говорить подробней в следующей части.

## Можно ли понимать код, не зная ни языка, ни программирования

Итак, мы с вами полны надежд, что код на Python легок для понимания. Ну я по крайней мере в этом уверен, а если вам прошлая часть этого еще не внушила, то сейчас будет еще несколько убедительных примеров.

Давайте смотреть на примеры кода и пробовать угадывать, что будет происходить при исполнении программы.

*Пример 1: переменные и списки*

In [1]:
bank_percent = 0.07
bank_percent_per_month = bank_percent / 12
initial_balance = [350000, 500000, 750000]

print("Проценты в конце первого месяца вклада:")
print(initial_balance[0] * bank_percent_per_month)

Проценты в конце первого месяца вклада:
2041.6666666666667


Варианты, что выведет код:
1. Результат умножения 350000 на 0.07/12
2. Результат умножения 500000 на 0.07/12
3. Результат умножения 750000 на 0.07/12
4. 350000 и 0.7/12 через пробел

Выбирайте свой вариант и давайте посмотрим, что происходит.

In [8]:
bank_percent = 0.07
bank_percent_per_month = bank_percent / 12
initial_balance = [350000, 500000, 750000]

print("Проценты в конце первого месяца вклада:")
print(initial_balance[0] * bank_percent_per_month)

Проценты в конце первого месяца вклада:
2041.6666666666667


С самого начала мы записываем в переменную bank_percent 0.07 (процент по вкладу в некотором банке), затем в bank_percent_per_month записываем одну двеннадцатую этой ставки (т.к. ставка годовая, а видимо это пример кода с рассчетом месячного дохода по вкладу). В переменную initial_balance записан целый список из нескольких чисел - возможный размер вклада. Затем выводится на печать строка, объясняющая, что будет выведено дальше, а после нее - доход от размещения на вкладе в банке 350000. С помощью [] мы получили один из элементов списка, а 0 - это номер элемента в списке. Нетрудно догадаться из примера, что элементы в списке нумеруются с нуля.

*Пример 2: ключевое слово def*

In [4]:
def calculate_first_month_profit(initial_balance, bank_percent):
    print(initial_balance * bank_percent / 12)

print(calculate_first_month_profit(350000, 0.05))
print(calculate_first_month_profit(500000, 0.06))

1458.3333333333333
None
2500.0
None


In [7]:
x = (2, 3, 4)

print(type(x))

<class 'tuple'>


Варианты, что выведет код:

1. Результат умножения 350000 на 0.05/12
1. Результат умножения 500000 на 0.06/12
1. И то и другое на отдельных строках

Убедимся, что все так, как вы и предположили:

In [10]:
def calculate_first_month_profit(initial_balance, bank_percent):
    return(initial_balance * bank_percent / 12)

print(calculate_first_month_profit(350000, 0.05))
print(calculate_first_month_profit(500000, 0.06))

1458.3333333333333
2500.0


В таком случае говорят, что мы определили свою функцию calculate_first_month_profit, которая принимает на вход два аргумента (initial_balance и bank_percent) и *возвращает* (return) результат. Результат вычисления значения функции при каких-то заданных значениях аргументов можно использовать на свое усмотрение - записать в какую-нибудь переменную, передать в print как в этом примере, или сразу проделать еще какие-то вычисления с возвращенным значением. Кстати, print - тоже функция, которой вы передаете то, что нужно вывести на печать, только ее вам определять не нужно - это стандартная функция Python.

*Пример 3: инструкция for*

In [ ]:
for i in range(0, 10):
    print(i)

Варианты, что делает код:
1. Уничтожает все данные на вашем компьютере
1. Запускает ракету в космос
1. Печатает букву "i"
1. Печатает букву "i" 10 раз
1. Записывает в переменную "i" по очереди числа от 0 до 10 и выводит их на печать

Выбирайте свой вариант и давайте посмотрим, что происходит.

In [8]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [9]:
print(type(range(1)))

<class 'range'>


И, в целом, логично. Просто читаем: for - пока, i in range(0, 10) - "i" в диапазоне от 0 до 10, print(i) - напечатать i. Так что правильным был последний вариант. Называется такая конструкция - цикл.

А вот если бы запускали ракету в космос, написали бы так:

In [6]:
for i in range(10, 0, -1):
    print(i)

10
9
8
7
6
5
4
3
2
1


Вы спросите, что это за -1? Ну, во-первых, вы наверняка уже сами догадываетесь, что значит этот аргумент функции range (да-да, range это тоже просто стандартная функция Python). Но, разумеется, обо всем этом всё равно будет рассказано по-порядку, но не прямо сейчас. Почему? Переходите скорее к продолжению урока, мы как раз это обсудим.

### Quiz

Попробуйте угадать, что напечает в процессе исполнения код, приведенный ниже. Этот тест покажет, что несмотря на общую интуитивность Python, бывают и примеры, которые заставят подумать человека, не программировавшего ранее. Но не волнуйтесь, вы можете проверить результат выполнения кода опытным путем - но перед проверкой обязательно решите для себя, в какой вариант больше верите, а если не угадали - подумайте, почему код работает так, как работает.

Вопрос 1.

a = 2
b = 3
a = b
b = a
print(a, b)

1. 2 3
1. 3 2
1. 2 2
1. 3 3

Вопрос 2.

a = 2
b = 3
a, b = b, a
print(a, b)

1. 2 3
1. 3 2
1. 2 2

Вопрос 3.

for i in ["Проблемы", "с доступом", "к joycasino?"]:
    print(i)
    
1. Фразу из одной из самых "доставучих" реклам в интернете, ставшей мемом, в одну строчку, все слова через пробелы
1. Все как в п.1, но перед "с доступом" и "к joycasion" будут переводы строк
1. Ничего не напечатает, т.к. вместо range написана какая-то чушь

Вопрос 4.

for i in range(10, 0, -1):
    print(i)
    if i == 1:
        print("Пуск!")
        
1. Только числа от 10 до 1 (включая 10 и 1, по одному на строку)
1. Числа от 10 до 1 (включая 10 и 1, по одному на строку) и на новой строке - "Пуск!"
1. Только числа от 10 до 0 (включая 10 и 0, по одному на строку)
1. Числа от 10 до 0 (включая 10 и 0, по одному на строку), а между 1 и 0 на отдельной строке - "Пуск!"

Вопрос 5.

for i in range(10, 0):
    print(i)
    
1. Числа от 0 до 10 в обратном порядке
1. Числа от 1 до 10 в обратном порядке
1. Ничего

Вопрос 6.

def calculate_first_month_profit(initial_balance, bank_percent):
    print(initial_balance * bank_percent / 12)

print(calculate_first_month_profit(350000, 0.05))
print(calculate_first_month_profit(500000, 0.06))

Подсказка: подумайте, возвращает ли результат вычислений функция calculate_first_month_profit

1. 1458.3333333333333

2500.0

2. 1458.3333333333333

None

2500.0

None

## Плюсы и минусы интуитивно понятного синтаксиса

В прошлых частях урока мы увидели, что можно неплохо понимать по коду на Python, что в нем происходит, даже не зная язык, и, возможно, никогда не программировав ранее. Чем это хорошо? Тем, что даже не зная толком Python мы можете найти в Интернете пример кода, делающего то, что вам нужно, немного подправить его и использовать.

Однако в этом есть и свои опасности. Во-первых, питоновский скрипт может делать и что-то плохое, например удалить все файл из папки, где он находится, или из какой-нибудь более важной. Поэтому не надо все подряд запускать на своем компьютере не глядя - если изучив код вы более-менее уверены, что он не представляет опасности - тогда можно попробовать запустить. Более того, лучше все равно его запустить в веб-интерпретаторе языка Python на каком-нибудь сайте - тогда скрипт будет выполняться не на вашем компьютере и вам ничего угрожать не будет.

Во-вторых, если всегда программировать только таким образом и не изучать возможности языка программирования и тонкости его использования, можно решать задачи неоптимально или вообще не знать, что что-то можно сделать с помощью Python. Кроме того, если раньше вы не сталкивались с программированием, наверняка куча абстракций из прошлой части (переменные, функции, циклы) несколько сбивает с толку в первый раз, и если вы познакомитесь со всем спокойно, последовательно и основательно - будет проще этими вещами оперировать.

Ну а теперь, когда все предупреждения сделаны, я предлагаю все-таки до следующего урока подождать с обстоятельным изложением и поразвивать умение понимать новый для вас код. Это умение поможет и эффективней проходить остаток курса, и сохранять живой азарт узнать, как же какой-то незнакомый код работает (и попробовать догадаться об этом), и программировать после этого курса. Ведь применяя программирование на практике вам придется решать именно ваши, уникальные задачи, и в точности их вы можете не встретить ни на каких курсах, а значит интуиция вам все равно пригодится.

## Кейс: поиск ошибки в коде

Представьте такую ситуацию: работаете вы, не имея никакого отношения к программированию, и тут к вам с горящими глазами приходит коллега, назовем его Вася, и рассказывает, что решил начать программировать на Python. Круто, модно, хайпово, еще и автоматизирует на работе все, что можно, получит премию и повышение.

Зная, как Вася часто что-то начинает и бросает, вы уже относитесь к этому с иронией, но он так хочет вам показать, как все просто на Питоне, что вы по дружбе соглашаетесь.

Придя к рабочему месту Васи, вы видите в одном окне открытый текстовый файл candidates.txt (видимо, с кандидатами на вакансию - Вася работает в HR), а вдругом окне - код:

In [13]:
%%writefile candidates.txt
Ivanov 1988
Petrov 1985
Sidorov 1990

Overwriting candidates.txt


In [16]:
"Petrov 1985".split(" ")

['Petrov', '1985']

In [17]:
input_file = open("candidates.txt")

min_year = 1985

for line in input_file.readlines():
    values = line.split(' ')
    values[1] = int(values[1])
    if values[1] <= min_year:
        print(line)

input_file.close()

Petrov 1985



Не имея ни малейшего представления, что этот код делает, и не встречавшись ранее ни с какими open, readline и split, вам остается только догадываться, что вам хочет продемонстрировать Вася. Кстати, как вы думаете, что выведет код?

1. Кандидатов, родившихся в 1985 году или раньше
2. Ничего

Вася начинает вам увлеченно рассказывать, как легко обрабатывать списки кандидатов на новую вакансию именно с помощью Питона, и уверяет, что его код сейчас отсеит кандидатов, родившихся позже 1985 года, и только их. Т.е. в его примере выведет только Петрова. Закатывая глаза с мыслью "Почему бы ему не использовать просто Эксель?!", вы терпеливо смотрите на результат работы скрипта, и...

In [24]:
input_file = open("candidates.txt")

min_year = 1985

for line in input_file.readlines():
    values = line.split(' ')
    values[1] = int(values[1])
    if values[1] < min_year:
        print(line)

input_file.close()

Скрипт не выводит ничего.

Пока Вася удивленно сверяет свой код с каким-то примером, который он скопировал из Интернета, в вас загорается азарт найти ошибку быстрее него. В конце-концов ну и пусть, что это какой-то непонятный код, вы же не просто так писали у себя в резюме "аналитический склад мышления". Логика очень проста: печатать что-то код должен тогда, когда вызывается print. print(line) приведет к печати какой-то строчки. Допустим это строчка из текстового файла (тогда input_file.readlines() в заголовке цикла видимо выдает список строчек из файла). Но на деле ведь не печатается ни одна строчка. Может быть что-то не так с if?

Почему код не вывел ничего?

1. Вася обратился к элементу номер 1 из списка values, а такого элемента в value нет
2. При чтении строки из файла в values[1] записывается год рождения кандидата, но Python не знает, что это число (т.к. по умолчанию - все считанное считается строкой). Вася забыл сделать строчку целым числом (int, сокращение от integer), поэтому фильтр по году и не работает.
3. Вася написал < min_year, а если посмотреть значение min_year - это 1985 год, т.е. Вася выводит только кандидатов СТАРШЕ чем 1985 года рождения, а таких в его файле нет. Нужно заменить знак < на <=.

Предложив Васе поправить строчку с if, вы тут же получили программу, работающую как надо. Внутри что-то торжествовало: столько неизвестных слов в коде, а вы исправили его работу! Но триумф длился не долго: Вася так впечатлился вашим умением находить ошибки в коде, что пообещал теперь обязательно приходить к вам за советом. Что ж, с эйчарами лучше дружить.

Обдумывая вечером эту ситуацию, вы поняли, что есть несколько интересных выводов про Питон:
1. Чтобы находить ошибки ("отлаживать код") можно просто постепенно разбирать код от того места, где проявляется ошибка, обращая внимание на то, что влияет на эту часть кода.
1. Можно пойти дальше и выводить некоторые из переменных на печать, чтобы увидеть, какие значения они принимают. Например, можно было всегда выводить массив values в коде Васи и просто воочию убедиться, что человек 1985 года рождения не выводится на печать, хотя год считался верно - значит точно проблема в строчке с if.
1. Можно использовать вывод на печать не только чтобы искать ошибки, но и чтобы понять, что происходит в чьем-то коде.

Поскольку ваш беспокойный коллега уже успел вам скинуть пример кода в почту, перед сном вы добавили в его код несколько вызовов print и попробовали разобраться, что происходит:

In [31]:
input_file = open("candidates.txt")

min_year = 1985

for line in input_file.readlines():
    print("line:", line)
    
    values = line.split(' ')
    print("values:", values)
    
    print("year:", values[1])
    
    values[1] = int(values[1])
    print("year:", values[1])
    
    print("if:", values[1] <= min_year)
    
    if values[1] <= min_year:
        print(line)
        
    print("-----")

input_file.close()

line: Ivanov 1988

values: ['Ivanov', '1988\n']
year: 1988

year: 1988
if: False
-----
line: Petrov 1985

values: ['Petrov', '1985\n']
year: 1985

year: 1985
if: True
Petrov 1985

-----
line: Sidorov 1990
values: ['Sidorov', '1990']
year: 1990
year: 1990
if: False
-----


## Кейс: автозаполнение документов

Эта история служит продолжением истории из прошлой части и завершает тему понимания и редактирования кода на Python. Прошло совсем немного времени и ваш коллега Вася снова вас настиг. В HR очень часто принимают каких-то людей на подряд и требуется заполнять в большом нудном договоре банковские реквизиты и паспортные данные. Вася придумал это автоматизировать (заставив людей присылать текстовый файл со своими данными в одинаковом формате и написав скрипт, который эти данные подставляет в документ в один клик), но ему нужна ваша помощь.

Вася утверждает, что для решения задачи потребуется много работать со строками и делится тем, что уже успел узнать.

Если у вас есть переменная, в которой записана строка, то у этой строки можно вызывать ряд методов, которые позволяют ее довольно хитрым образом обрабатывать.

Например, разбивать по разделительному знаку:

In [32]:
candidate = "Vasya,1985,1234 56789,Khamovniki,2005"

print(candidate.split(','))

['Vasya', '1985', '1234 56789', 'Khamovniki', '2005']


И заменять какие-то части строки:

In [33]:
text = "Agreement with $EMPLOYEE$"

print(text.replace('$EMPLOYEE$', 'Vasya'))

Agreement with Vasya


Кстати, нашел эти примеры Вася просто написав в поисковике "как в Python разбить строку по разделителю" и "как в Python заменить часть строки", так что оказывается, для программирования иногда достаточно поисковика и умения сформулировать, что вам нужно сделать.

Также Вася уже сообразил, что можно завести в текстовом файле шаблон договора, где вместо конкретных имен и данных будут какие-то слова, которые будет заменять скрипт. Например, вместо имени работника - $EMPLOYEE$, вместо серии и номера паспорта - $ID_NUM$ и так далее. Более того, Вася настолько обленился, что из старого договра, который до трудоустройства составили с ним, почти сделал такой шаблон тоже с помощью питона (и даже в паре мест написал для вас комментарии в коде). Про кнопку "найти и заменить" в Ворде Вася, похоже, не слышал.

In [ ]:
input_file = open("Vasya_agreement.txt", 'r') # r - for reading
text = input_file.read()
input_file.close()

text.replace("Vasya", "$EMPLOYEE$")
text.replace("1985", "$BORN$")
text.replace("1234 56789", "$ID_NUM$")
text.replace("Khamovniki", "$ID_REGION$")
text.replace("2005", "$ID_DATE$")

output_file = open("Template.txt", "w") # w - for writing
output_file.write(text)
output_file.close()

Проблема с кодом в том, что почему-то итоговый шаблон договора получается в точности таким же документом (без замен). Похоже, снова нужно искать ошибку.

Как вы думаете, почему текст остался таким же?

1. Вася опечатался в своих данных, когда писал код, заменяющий их
1. Вася вообще никуда не сохранял текст договора после замен
1. Файл не считался изначально

Разумеется, из примеров выше следует, что text.replace возвращает текст после замены, а возвращенное значение еще надо куда-то сохранить. Код Васи же написан так, как будто text.replace делает замену прямо в переменной text и полученная строчка остается там. После правки кода получилось:


In [ ]:
input_file = open("Vasya_agreement.txt", 'r') # r - for reading
text = input_file.read()
input_file.close()

text = text.replace("Vasya", "$EMPLOYEE$")
text = text.replace("1985", "$BORN$")
text = text.replace("1234 56789", "$ID_NUM$")
text = text.replace("Khamovniki", "$ID_REGION$")
text = text.replace("2005", "$ID_DATE$")

output_file = open("Template.txt", "w") # w - for writing
output_file.write(text)
output_file.close()

Вы снова все починили, и теперь Вася просит вас дописать скрипт, который возьмет файл с данными разных людей и сгенерирует по договору на каждого. Ну что ж, вам по крайней мере понятно, что задача решается копированием Васиного кода и небольшими исправлениями. Еще потребуется цикл вроде того, что был в прошлом вопросе от Васе, и сохранять новые текстовые файлы тоже видимо понадобится в цикле. Можете для примера набросать какой-нибудь текстовый документ, где нужно будет заменить некоторые поля на реальные данные людей и попробовать сделать аналогичную задачу.

## Чему мы научились и что будет дальше

1. Не боимся страшного нового кода, даже если чего-то не понимаем. Код в целом логичен
1. Можно находить примеры кода, править их по мелочи и так учиться что-то делать
1. Если постоянно делать только так, будет страшно писать код с чистого листа, поэтому дальше мы последовательно поговорим в деталях про различные возможности языка